### PR TITLE
🚀 Tenary Weight and DoReFa-Net in TensorFlow (TensorLayer)

### PR DESCRIPTION
**Pull request by XJTUWYD**
I add two compress strategies tenary weight network  and dorefa-net into tensorlayer,  did two experiments to compare the accuracy of different compress strategies based on mnist and cifar-10.
the result of the experiment is below:

|             	| BinaryNet      	| Tenary Weight      	| DoReFa-Net      	|
|-------------	|---------------	|---------------	|---------------	|
| MNIST 	| 98.86% 	| 99.27% 	| 98.89% 	|
| CIFAR10  	| 41.1% 	| 80.6% 	| 81.1% 	|

related issue: https://github.com/tensorlayer/tensorlayer/issues/416

### CODE REVIEW

File: tensorlayer/layers/binary.py [hunk 1]
```diff
@@ -2,10 +2,18 @@
 from .core import *
 from .. import _logging as logging
 import tensorflow as tf
+from tensorflow.python.framework import ops
+
+bitW = 1
```
Comment (luomai) [inline tensorlayer/layers/binary.py]: can you avoid using global variables to setup the parameters?

File: tensorlayer/layers/binary.py [hunk 2]
```diff
@@ -24,6 +32,71 @@ def quantize(x):
         return tf.sign(x)
 
 
+def quantize_dorefa(x, k):
+    G = tf.get_default_graph()
+    n = float(2**k - 1)
+    with G.gradient_override_map({"Round": "Identity"}):
+        return tf.round(x * n) / n
+
+def fw(x, force_quantization=False):
```
Comment (luomai) [inline tensorlayer/layers/binary.py]: can we use a more meaningful names? If this function is used only internally, we better to start the name with a underscore.

File: tensorlayer/layers/binary.py [hunk 3]
```diff
@@ -224,6 +297,409 @@ def __init__(
             self.all_params.append(W)
 
 
+class TenaryDenseLayer(Layer):
```
Comment (luomai) [inline tensorlayer/layers/binary.py]: we usually recommend to add a few unit tests to layers.

File: example/tutorial_tenarycifar10_tfrecord.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+Reading images from disk and distorting them can use a non-trivial amount
+of processing time. To prevent these operations from slowing down training,
+we run them inside 16 separate threads which continuously fill a TensorFlow queue.
+
+"""
+
+import os, time
+import tensorflow as tf
+import tensorlayer as tl
+from tensorlayer.layers import *
```
Comment (luomai) [inline example/tutorial_tenarycifar10_tfrecord.py]: We prefer to avoiding "import *" in new committed files.

File: tensorlayer/layers/binary.py [hunk 5]
```diff
@@ -2,10 +2,18 @@
 from .core import *
```
Comment (lgarithm) [inline tensorlayer/layers/binary.py]: don't use `import *` for new code.

**Comment by luomai** [comment]
@zsdonghao as we are having a large number of examples, shall we actually start organizing the examples into different folders?

File: example/tutorial_dorefacifar10_tfrecord.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+            net = tl.layers.InputLayer(x_crop, name='input')
+            net = tl.layers.Conv2d(net, 64, (5, 5), (1, 1), act=tf.nn.relu, padding='SAME', W_init=W_init, name='cnn1')
+            net = tl.layers.MaxPool2d(net, (3, 3), (2, 2), padding='SAME', name='pool1')
+            net = tl.layers.LocalResponseNormLayer(net, depth_radius=4, bias=1.0, alpha=0.001 / 9.0, beta=0.75, name='norm1')
+            net = tl.layers.DorefaConv2d(net, 1, 3, 64, (5, 5), (1, 1), act=tf.nn.relu, padding='SAME', W_init=W_init, name='cnn2')
+            net = tl.layers.LocalResponseNormLayer(net, depth_radius=4, bias=1.0, alpha=0.001 / 9.0, beta=0.75, name='norm2')
+            net = tl.layers.MaxPool2d(net, (3, 3), (2, 2), padding='SAME', name='pool2')
+            net = tl.layers.FlattenLayer(net, name='flatten')  # output: (batch_size, 2304)
+            net = tl.layers.DorefaDenseLayer(net, 1, 3, n_units=384, act=tf.nn.relu, W_init=W_init2, b_init=b_init2, name='d1relu')  # output: (batch_size, 384)
+            net = tl.layers.DorefaDenseLayer(net, 1, 3, n_units=192, act=tf.nn.relu, W_init=W_init2, b_init=b_init2, name='d2relu')  # output: (batch_size, 192)
```
[the remark this answers is not in the document]
Comment (luomai) [reply example/tutorial_dorefacifar10_tfrecord.py]: @lgarithm  what does this issue imply?

File: example/tutorial_binarycifar10_tfrecord.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
+#! /usr/bin/python
+# -*- coding: utf-8 -*-
+"""Reimplementation of the TensorFlow official CIFAR-10 CNN tutorials.
```
Comment (zsdonghao) [inline example/tutorial_binarycifar10_tfrecord.py]: we can remove this comment.

File: example/tutorial_binarycifar10_tfrecord.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    label = tf.cast(features['label'], tf.int32)
+    return img, label
+
+
+## Save data into TFRecord files
+data_to_tfrecord(images=X_train, labels=y_train, filename="train.cifar10")
+data_to_tfrecord(images=X_test, labels=y_test, filename="test.cifar10")
+
+## Example to visualize data
+# img, label = read_and_decode("train.cifar10", None)
```
Comment (zsdonghao) [inline example/tutorial_binarycifar10_tfrecord.py]: we can remove this comment.

File: example/tutorial_dorefacifar10_tfrecord.py [hunk 9]
```diff
@@ -0,0 +1,319 @@
+#! /usr/bin/python
+# -*- coding: utf-8 -*-
+"""Reimplementation of the TensorFlow official CIFAR-10 CNN tutorials.
```
Comment (zsdonghao) [inline example/tutorial_dorefacifar10_tfrecord.py]: we can remove this comment.

File: example/tutorial_dorefacifar10_tfrecord.py [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+
+    label = tf.cast(features['label'], tf.int32)
+    return img, label
+
+
+## Save data into TFRecord files
+data_to_tfrecord(images=X_train, labels=y_train, filename="train.cifar10")
+data_to_tfrecord(images=X_test, labels=y_test, filename="test.cifar10")
+
+## Example to visualize data
```
Comment (zsdonghao) [inline example/tutorial_dorefacifar10_tfrecord.py]: we can remove this comment.

File: example/tutorial_tenarycifar10_tfrecord.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
+#! /usr/bin/python
+# -*- coding: utf-8 -*-
+"""Reimplementation of the TensorFlow official CIFAR-10 CNN tutorials.
```
Comment (zsdonghao) [inline example/tutorial_tenarycifar10_tfrecord.py]: we can remove this comment.

File: example/tutorial_tenarycifar10_tfrecord.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    label = tf.cast(features['label'], tf.int32)
+    return img, label
+
+
+## Save data into TFRecord files
+data_to_tfrecord(images=X_train, labels=y_train, filename="train.cifar10")
+data_to_tfrecord(images=X_test, labels=y_test, filename="test.cifar10")
+
+## Example to visualize data
+# img, label = read_and_decode("train.cifar10", None)
```
Comment (zsdonghao) [inline example/tutorial_tenarycifar10_tfrecord.py]: we can remove this comment.

File: tensorlayer/layers/binary.py [hunk 13]
```diff
@@ -224,6 +295,424 @@ def __init__(
...
+    use_cudnn_on_gpu : bool
+        Default is False.
+    data_format : str
+        "NHWC" or "NCHW", default is "NHWC".
+    name : str
+        A unique layer name.
+
+    Examples
+    ---------
+    >>> net = tl.layers.InputLayer(x, name='input')
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: the example here should be modified.

File: tensorlayer/layers/binary.py [hunk 14]
```diff
@@ -224,6 +295,424 @@ def __init__(
...
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: the example here should be modified

File: tensorlayer/layers/binary.py [hunk 15]
```diff
@@ -224,6 +295,424 @@ def __init__(
...
+            b_init_args = {}
+
+        if use_gemm:
+            raise Exception("TODO. The current version use tf.matmul for inferencing.")
+
+        Layer.__init__(self, prev_layer=prev_layer, name=name)
+        self.inputs = prev_layer.outputs
+        if act is None:
+            act = tf.identity
+        logging.info("BinaryConv2d %s: n_filter:%d filter_size:%s strides:%s pad:%s act:%s" % (self.name, n_filter, str(filter_size), str(strides), padding,
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: change  BinaryConv2d to DorefaConv2d

File: tensorlayer/layers/binary.py [hunk 16]
```diff
@@ -224,6 +295,424 @@ def __init__(
...
+        self.inputs = prev_layer.outputs
+        if self.inputs.get_shape().ndims != 2:
+            raise Exception("The input dimension must be rank 2, please reshape or flatten it")
+
+        if use_gemm:
+            raise Exception("TODO. The current version use tf.matmul for inferencing.")
+
+        n_in = int(self.inputs.get_shape()[-1])
+        self.n_units = n_units
+        logging.info("BinaryDenseLayer  %s: %d %s" % (self.name, self.n_units, act.__name__))
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: change BinaryDenseLayer  to DorefaDenseLayer

File: tensorlayer/layers/binary.py [hunk 17]
```diff
@@ -224,6 +295,424 @@ def __init__(
...
+            b_init_args = {}
+
+        if use_gemm:
+            raise Exception("TODO. The current version use tf.matmul for inferencing.")
+
+        Layer.__init__(self, prev_layer=prev_layer, name=name)
+        self.inputs = prev_layer.outputs
+        if act is None:
+            act = tf.identity
+        logging.info("BinaryConv2d %s: n_filter:%d filter_size:%s strides:%s pad:%s act:%s" % (self.name, n_filter, str(filter_size), str(strides), padding,
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: change BinaryConv2d  to TenaryConv2d

File: tensorlayer/layers/binary.py [hunk 18]
```diff
@@ -224,6 +295,424 @@ def __init__(
...
+        self.inputs = prev_layer.outputs
+        if self.inputs.get_shape().ndims != 2:
+            raise Exception("The input dimension must be rank 2, please reshape or flatten it")
+
+        if use_gemm:
+            raise Exception("TODO. The current version use tf.matmul for inferencing.")
+
+        n_in = int(self.inputs.get_shape()[-1])
+        self.n_units = n_units
+        logging.info("BinaryDenseLayer  %s: %d %s" % (self.name, self.n_units, act.__name__))
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: change BinaryDenseLayer  to TenaryDenseLayer

File: tensorlayer/layers/binary.py [hunk 19]
```diff
@@ -24,6 +28,73 @@ def quantize(x):
         return tf.sign(x)
 
 
+def _quantize_dorefa(x, k):
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: need comment

File: tensorlayer/layers/binary.py [hunk 20]
```diff
@@ -24,6 +28,73 @@ def quantize(x):
...
+    G = tf.get_default_graph()
+    n = float(2**k - 1)
+    with G.gradient_override_map({"Round": "Identity"}):
+        return tf.round(x * n) / n
+
+
+def _quantize_weight(x, bitW, force_quantization=False):
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: need comment

File: tensorlayer/layers/binary.py [hunk 21]
```diff
@@ -24,6 +28,73 @@ def quantize(x):
...
+        return x
+    if bitW == 1:  # BWN
+        with G.gradient_override_map({"Sign": "Identity"}):
+            E = tf.stop_gradient(tf.reduce_mean(tf.abs(x)))
+            return tf.sign(x / E) * E
+    x = tf.clip_by_value(x * 0.5 + 0.5, 0.0, 1.0)  # it seems as though most weights are within -1 to 1 region anyways
+    return 2 * _quantize_dorefa(x, bitW) - 1
+
+
+def _quantize_active(x, bitA):
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: need comment

File: tensorlayer/layers/binary.py [hunk 22]
```diff
@@ -24,6 +28,73 @@ def quantize(x):
...
+    if bitA == 32:
+        return x
+    return _quantize_dorefa(x, bitA)
+
+
+def _cabs(x):
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: need comment

File: tensorlayer/layers/binary.py [hunk 23]
```diff
@@ -24,6 +28,73 @@ def quantize(x):
...
+    return tf.minimum(1.0, tf.abs(x), name='cabs')
+
+
+def _compute_threshold(x):
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: need comment

File: tensorlayer/layers/binary.py [hunk 24]
```diff
@@ -24,6 +28,73 @@ def quantize(x):
...
+    ref: https://github.com/XJTUWYD/TWN
+    getting the threshold
+    '''
+    x_sum = tf.reduce_sum(tf.abs(x), reduction_indices=None, keep_dims=False, name=None)
+    threshold = tf.div(x_sum, tf.cast(tf.size(x), tf.float32), name=None)
+    threshold = tf.multiply(0.7, threshold, name=None)
+    return threshold
+
+
+def _compute_alpha(x):
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: need comment

File: tensorlayer/layers/binary.py [hunk 25]
```diff
@@ -24,6 +28,73 @@ def quantize(x):
...
+    alpha_array = tf.add(alpha1_temp1, alpha1_temp2, name=None)
+    alpha_array_abs = tf.abs(alpha_array)
+    alpha_array_abs1 = tf.where(tf.greater(alpha_array_abs, 0), tf.ones_like(alpha_array_abs, tf.float32), tf.zeros_like(alpha_array_abs, tf.float32))
+    alpha_sum = tf.reduce_sum(alpha_array_abs)
+    n = tf.reduce_sum(alpha_array_abs1)
+    alpha = tf.div(alpha_sum, n)
+    return alpha
+
+
+def _tenary_opration(x):
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: need comment

File: example/tutorial_dorefanet_mnist_cnn.py [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+x = tf.placeholder(tf.float32, shape=[batch_size, 28, 28, 1])
+y_ = tf.placeholder(tf.int64, shape=[batch_size])
+
+
+def model(x, is_train=True, reuse=False):
+    # In BNN, all the layers inputs are binary, with the exception of the first layer.
+    # ref: https://github.com/itayhubara/BinaryNet.tf/blob/master/models/BNN_cifar10.py
+    with tf.variable_scope("binarynet", reuse=reuse):
+        net = tl.layers.InputLayer(x, name='input')
+        net = tl.layers.DorefaConv2d(net, 32, (5, 5), (1, 1), padding='SAME', b_init=None, name='bcnn1')
```
Comment (lgarithm) [inline example/tutorial_dorefanet_mnist_cnn.py]: How does `DorefaConv2d` know that `32` is for `n_filter` but not for `bitW` or `bitA`?

Comment (lgarithm) [reply example/tutorial_dorefanet_mnist_cnn.py]: it crashed with
```
$ python3 example/tutorial_dorefanet_mnist_cnn.py 
[TL] Load or Download MNIST > data/mnist
[TL] data/mnist/train-images-idx3-ubyte.gz
[TL] data/mnist/t10k-images-idx3-ubyte.gz
2018-03-22 21:30:07.666529: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2
[TL] InputLayer  binarynet/input: (128, 28, 28, 1)
Traceback (most recent call last):
  File "example/tutorial_dorefanet_mnist_cnn.py", line 48, in <module>
    net_train = model(x, is_train=True, reuse=False)
  File "example/tutorial_dorefanet_mnist_cnn.py", line 25, in model
    net = tl.layers.DorefaConv2d(net, 32, (5, 5), (1, 1), padding='SAME', b_init=None, name='bcnn1')
  File "/home/vagrant/.local/lib/python3.6/site-packages/tensorlayer/layers/binary.py", line 687, in __init__
    act.__name__))
TypeError: %d format: a number is required, not tuple
```

**Comment by lgarithm** [comment]
Since our functions have long argument lists, and many of them have default values, I suggest that we always use keyword style argument passing (except for small helper functions that 1,2 or 3 arguments).

File: example/tutorial_dorefanet_mnist_cnn.py (hunk 26 above)
Comment (XJTUWYD) [reply example/tutorial_dorefanet_mnist_cnn.py]: I forgot it and I have fixed the bug

File: tensorlayer/layers/binary.py [hunk 27]
```diff
@@ -587,8 +588,8 @@ def __init__(
 
 
 class DorefaConv2d(Layer):
-    """
-    The :class:`BinaryConv2d` class is a 2D binary CNN layer, which weights are either -1 or 1 while inferencing.
+    """The :class:`DorefaConv2d` class is a binary fully connected layer, which weights are 'bitW' bits and the input of the previous layer
```
Comment (zsdonghao) [inline tensorlayer/layers/binary.py]: should be convolutional layer

**Comment by XJTUWYD** [comment]
bnn is  a excellent work in the compression of neuron network but it can not get a satisfied accuracy on relative large datasets, in order to solve the problem, tenary weight networks and dorefa were put forward. I add 4 apis for tensorlayer, Tenary Denselayer, TenaryConv2d, DorefaDenselayer, and DorefaConv2d . I perform 6 experiment based on mnist and cifar10，the details are in thr titorials.
finally thank you very much for the help of HaoDong, LuoMai, and Igarithm.

**Comment by zsdonghao** [comment]
@XJTUWYD Hi, i suggest you to add this comment into this issue https://github.com/tensorlayer/tensorlayer/issues/416, more people can see it.

**Pull request merged**